components/builder-worker/src/server.rs

```diff
@@ -145,6 +149,16 @@ impl Server {
                 }
                 fe_msg = false;
             }
+
+            // If we do not receive a new job or have a job status from the runner,
+            // then send out a heartbeat.
+            if !runner_msg && !fe_msg {
+                match self.state {
+                    State::Ready => self.heartbeat.set_state(jobsrv::WorkerState::Ready),
+                    State::Busy => self.heartbeat.set_state(jobsrv::WorkerState::Busy),
+                }
+                self.send_heartbeat()?;
```

Comment:
I'm thinking this entire PR could probably be simplified quite a bit if you left the heartbeating be as is and here just send an empty message to the jobsrv.